crates/bevy_image/src/image.rs

```diff
@@ -1504,14 +1554,180 @@ impl Image {
+    pub fn format(&self) -> TextureFormat {
+        self.format
+    }
+}
+impl<'a> Iterator for RectBytes<'a> {
+    type Item = &'a [u8];
+    fn next(&mut self) -> Option<&'a [u8]> {
+        self.data.next().map(|s| &s[self.range.clone()])
+    }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len(), Some(self.len()))
+    }
+}
+impl<'a> DoubleEndedIterator for RectBytes<'a> {
+    fn next_back(&mut self) -> Option<&'a [u8]> {
+        self.data.next_back().map(|s| &s[self.range.clone()])
+    }
+}
+impl<'a> ExactSizeIterator for RectBytes<'a> {
+    fn len(&self) -> usize {
+        self.data.len()
+    }
+}
+impl<'a> FusedIterator for RectBytes<'a> {}
+
+/// Iterator returned by [`Image::rect_bytes_mut`].
+pub struct RectBytesMut<'a> {
+    data: slice::ChunksExactMut<'a, u8>,
+    range: Range<usize>,
+    format: TextureFormat,
+}
+impl RectBytesMut<'_> {
+    /// Gets the texture format for the rect.
+    pub fn format(&self) -> TextureFormat {
+        self.format
+    }
+
+    /// Directly copies the bytes from another rect to this one.
+    ///
+    /// Will fail if the rects are different sizes or formats. Completely
+    /// ignores transparency and will directly overwrite the bytes.
+    pub fn copy_from(self, other: RectBytes<'_>) -> Result<(), TextureCopyError> {
+        let dest = UVec2::new(self.range.len() as u32, self.data.len() as u32);
+        let src = UVec2::new(other.range.len() as u32, other.data.len() as u32);
+        if dest != src {
+            return Err(TextureCopyError::SizeMismatch { dest, src });
+        }
+
+        if self.format != other.format {
+            return Err(TextureCopyError::FormatMismatch {
+                src: self.format,
+                dest: other.format,
+            });
+        }
+
+        for (dest, src) in self.zip(other.clone()) {
+            dest.copy_from_slice(src);
+        }
+        Ok(())
+    }
+
+    /// Copies the pixels from another rect to this one by performing color conversion.
+    ///
+    /// Will fail if the rects are different sizes or formats. Completely
+    /// ignores transparency and will directly overwrite the colors.
+    pub fn translate_from(self, other: RectBytes<'_>) -> Result<(), TextureCopyError> {
+        if self.format == other.format {
+            return self.copy_from(other);
+        }
+
+        let dest_len = self.format.pixel_size();
+        let src_len = other.format.pixel_size();
+
+        let dest = UVec2::new((self.range.len() / dest_len) as u32, self.data.len() as u32);
+        let src = UVec2::new(
+            (other.range.len() / src_len) as u32,
+            other.data.len() as u32,
+        );
+
+        if dest != src {
+            return Err(TextureCopyError::SizeMismatch { src, dest });
+        }
+
+        if !Image::supported_color_format(self.format)
+            || !Image::supported_color_format(other.format)
+        {
+            return Err(TextureCopyError::FormatMismatch {
```

Comment:
I think this error does not quite fit for this case either. At this point in the code the formats will always "mismatch".

Reply:
They won't match, but it's only if they can't be converted that the conversion will fail. Unless you're referring to something else.

Reply:
I think it is ultimately fine, I just got confused by the name of the error variant and the error message.